## GRIB: getting latitudes, longitudes and values (reduced Gaussian grid)

In this example we will work with GRIB data on a reduced Gaussian GRIB.

In [1]:
import earthkit.data as ekd

In [2]:
ds = ekd.from_source("sample", "tuv_pl_rgg_O32.grib").sel(param="t")
ds.ls()

tuv_pl_rgg_O32.grib:   0%|          | 0.00/190k [00:00<?, ?B/s]

centre shortName    typeOfLevel  level  dataDate  dataTime stepRange  \
0   ecmf         t  isobaricInhPa   1000  20180801      1200         0   
1   ecmf         t  isobaricInhPa    850  20180801      1200         0   
2   ecmf         t  isobaricInhPa    700  20180801      1200         0   
3   ecmf         t  isobaricInhPa    500  20180801      1200         0   
4   ecmf         t  isobaricInhPa    400  20180801      1200         0   
5   ecmf         t  isobaricInhPa    300  20180801      1200         0   

  dataType  number    gridType  
0       an       0  reduced_gg  
1       an       0  reduced_gg  
2       an       0  reduced_gg  
3       an       0  reduced_gg  
4       an       0  reduced_gg  
5       an       0  reduced_gg

In the data each field is defined on an O32 octahedral reduced Gaussian grid. These grids cannot be represented as a 2D array so the field shape is always 1 dimensional:

In [3]:
ds[0].shape

(5248,)

### Using the data() method

#### Fields

In [4]:
llv = ds[0].data()
type(llv), llv.shape

(numpy.ndarray, (3, 5248))

Here *llv[0]* contains the latitudes, *llv[1]* the longitudes, while *llv[2]* the values. Each of these arrays has the same shape that of the field *(5248,)*).

In [5]:
print(llv[0].shape)
# first latitude
llv[0, 0]

(5248,)


87.86379883923263

In [6]:
print(llv[1].shape)
# first longitude
llv[1, 0]

(5248,)


0.0

In [7]:
print(llv[2].shape)
# first value
llv[2, 0]

(5248,)


271.55796813964844

#### FieldLists

In [8]:
llv = ds.data()
type(llv), llv.shape

(numpy.ndarray, (8, 5248))

Latitudes can be accessed as *llv[0]*.

In [9]:
print(llv[0].shape)
# first latitude
llv[0, 0]

(5248,)


87.86379883923263

Longitudes can be accessed as *llv[1]*.

In [10]:
print(llv[1].shape)
# first longitude
llv[1, 0]

(5248,)


0.0

The field values can be accessed as *llv[2:]*.

In [11]:
print(llv[2:].shape)

# first value in first field
print(llv[2, 0])

# first value from all the fields
print(llv[2:, 0])

(6, 5248)
271.55796813964844
[271.55796814 277.64489746 272.19042969 255.98562622 243.70146179
 228.54637146]


### Using the to_latlon() and to_numpy() methods

#### Fields

In [12]:
ll = ds[0].to_latlon()
for k, v in ll.items():
    print(f"{k} shape={v.shape}")

lat shape=(5248,)
lon shape=(5248,)


In [13]:
v = ds[0].to_numpy()
v.shape

(5248,)

By default both methods keep the field's shape, but we can use the *flatten* keyword to get 1D arrays:

In [14]:
ll = ds[0].to_latlon(flatten=True)
v = ds[0].to_numpy(flatten=True)
print(ll["lat"].shape)
print(ll["lon"].shape)
print(v.shape)

(5248,)
(5248,)
(5248,)


#### FieldLists

In [15]:
ll = ds.to_latlon()
for k, v in ll.items():
    print(f"{k} shape={v.shape}")

lat shape=(5248,)
lon shape=(5248,)


In [16]:
v = ds.to_numpy()
v.shape

(6, 5248)

E.g. the first value from each field can be extracted as:

In [17]:
v[:,0]

array([271.55796814, 277.64489746, 272.19042969, 255.98562622,
       243.70146179, 228.54637146])

### Specifying the array type

For all the methods above we can set the array type with the *dtype* keyword both for fields and fieldlists:

In [18]:
import numpy as np
v = ds.to_numpy(dtype=np.float32)
v[:,0]

array([271.55798, 277.6449 , 272.19043, 255.98563, 243.70146, 228.54637],
      dtype=float32)

In [19]:
llv = ds.data(dtype=np.float32)

In [20]:
llv[:,0]

array([ 87.8638 ,   0.     , 271.55798, 277.6449 , 272.19043, 255.98563,
       243.70146, 228.54637], dtype=float32)